Replace invite_by_admins_only with invite_to_realm_policy and also add option of moderators and full members for this setting.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
- First commit is to just convert the bool `invite_by_admins_only` to an integer field `invite_to_realm_policy`. Added moderators and full members option in separate commit as this commit also included frontend changes and migrations, so it will be easy to review.
- Second commit adds helper `can_invite_others_to_realm` similar to `can_create_streams` which will be used in further commits after we add other options to `invite_to_realm_policy`
- Third commit adds the moderators and full members option to `invite_to_realm_policy` by using `COMMON_POLICY_TYPES`.
- Fourth commit is to send `can_invite_others_to_realm` to the clients as in `page_params` dict.
- Fifth commit fixes the bug of incorrectly showing invite link even when user is not allowed to invite others.
- Sixth commit fixes the tests for `create_stream_policy` and `invite_to_stream_policy` which was missed in previous commits.

**Testing plan:** <!-- How have you tested? --> Modified and added tests.


<!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented:
This is ready for review. Have added a commit for solving the bug and there is detailed explanation in the commit message.

Also @timabbott I think instead of having separate tests for `can_create_streams` and other functions using `has_permission`, we can have a single helper function which will be called for different policies, this will be helpful as we have planned to have other policies to use `COMMON_POLICY_TYPES` in future. I just wanted to ask what is the correct place of that helper function, is `test_classes.py` fine and the functions calling it will be in different files according to policies.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2939,12 +2939,16 @@ paths:
+                                        * 1 = Members only
+                                        * 2 = Administrators only
+
+                                        **Changes**: New in Zulip 4.0 (feature level 47).
```
We should add here what the old field was.  E.g. add a new sentence.  "Previously, this was encoded as the boolean field `invite_by_admins_only`".

timabbott commented on zerver/models.py:
```diff
@@ -272,6 +262,8 @@ class Realm(models.Model):
         POLICY_MODERATORS_ONLY,
     ]
 
+    invite_to_realm_policy: int = models.PositiveSmallIntegerField(default=POLICY_MEMBERS_ONLY)
```
Maybe we should sort this alphabetically among this block (i.e. after `create_stream_policy`), and also add a descriptive comment?

timabbott commented on zerver/views/invite.py:
```diff
@@ -39,11 +39,16 @@ def invite_users_backend(
+            )
+        if user_profile.realm.invite_to_realm_policy == Realm.POLICY_FULL_MEMBERS_ONLY:
+            return json_error(_("Your account is too new to invite others to this organization."))
+        # Guest case will be handled by require_member_or_admin decorator.
```
This mostly works, but I feel like this structure of having this block of error-handling checks feels really fragile, and we should figure out how to refactor it.  One thought is that we should use common error messages -- e.g. "Must be an organization administrator" or "Must be a moderator or organization administrator" -- that don't name the specific thing you're trying to do (To simplify translations), though it'd be nice to also encode that we're 

Probably fine to leave this as a future problem, but it does seem important.

That said, I think even for a temporary state, we should have a catch-all clause (even a `raise AssertionError()`) -- here if we added a new policy value, and made `user_profile.can_invite_others_to_realm` return false, this structure would cause that to get through.

timabbott commented:
> I just wanted to ask what is the correct place of that helper function, is test_classes.py fine and the functions calling it will be in different files according to policies.

I think `ZulipTestCase` is an OK place for that sort of thing.

sahil839 commented:
@timabbott updated the PR with -
- Sorted the policies in `models.py` and also added a descriptive comment.
- Added about the old field in `zulip.yaml` in a similar way we do for `stream_post_policy`.

Regarding the third point to raise assertion error, It seems you meant the following changes - 
```
--- a/zerver/views/invite.py
+++ b/zerver/views/invite.py
@@ -49,6 +49,8 @@ def invite_users_backend(
         if user_profile.realm.invite_to_realm_policy == Realm.POLICY_FULL_MEMBERS_ONLY:
             return json_error(_("Your account is too new to invite others to this organization."))
         # Guest case will be handled by require_member_or_admin decorator.
+        if user_profile.realm.invite_to_realm_policy != Realm.POLICY_MEMBERS_ONLY
+            raise AssertionError(_("Invalid value of invite_to_realm_policy."))
     if invite_as not in PreregistrationUser.INVITE_AS.values():
         return json_error(_("Must be invited as an valid type of user"))
     check_if_owner_required(invite_as, user_profile)
```
And the `has_permission` code is as follows, where we cannot generate a condition where `can_invite_others_to_realm` return false when we set the `invite_to_realm_policy` as some invalid value like 10 as we also assert at last that the policy value is of full members only, and for other cases where we return false before that we do so for some certain policy values for which we already handle errors and for guest users, it is handled using decorator so code won't come at this point  - 
```
def has_permission(self, policy_name: str) -> bool:
        if policy_name not in [
            "create_stream_policy",
            "invite_to_stream_policy",
            "invite_to_realm_policy",
        ]:
            raise AssertionError("Invalid policy")

        if self.is_realm_admin:
            return True

        policy_value = getattr(self.realm, policy_name)
        if policy_value == Realm.POLICY_ADMINS_ONLY:
            return False

        if self.is_moderator:
            return True

        if policy_value == Realm.POLICY_MODERATORS_ONLY:
            return False

        if self.is_guest:
            return False

        if policy_value == Realm.POLICY_MEMBERS_ONLY:
            return True

        assert policy_value == Realm.POLICY_FULL_MEMBERS_ONLY
        return not self.is_provisional_member
```

sahil839 commented:
And regarding the deduplication of tests, I will add those in a different PR as this PR is already large enough and for test of this policy I will make a separate PR or do changes in this only depending on which is merged first.

sahil839 commented:
> And regarding the deduplication of tests, I will add those in a different PR as this PR is already large enough and for test of this policy I will make a separate PR or do changes in this only depending on which is merged first.

Opened #18015 for this.

timabbott commented:
@sahil839 can you do a rebase on this?  I just merged two larger PRs with API changelog + migrations, and this is next in my sequence of API migrations to try to merge.

sahil839 commented:
@timabbott rebased and also changed the tests here to use helper function merged in #18015.

sahil839 commented:
@timabbott fixed the import issue, ready for review.

timabbott commented:
For https://github.com/zulip/zulip/pull/17990#issuecomment-814312780, I realize we don't currently have a bug, but the code structure is fragile -- if we added a new possible value, things could just drop through with the value even with a validation failure.  

I added 2c9a1e085cc86aef6f073db306e998ece6ab0bc5 to defense against that sort of bug for now.  

However, I think what I actually want to do is replace the `has_permission` blocks of error-reporting logic with something like

```
if not user_profile.can_create_streams():
     raise JsonableError(_("Insufficient permission"))
```

We expect the frontend to be doing enforcement about what options to show anyway, so explaining the details for why the user isn't permitted in the error message isn't important (since we don't expect to actually show these error messages to users often; it'll mainly be via the API).

Want to do that change?  It'd let us delete code and delete strings for translators while eliminating this class of bug; the only downside is that we'll need to change a bunch of tests that previously verified these more descriptive error messages.

sahil839 commented:
Ohh, I did not realize that the `raise AssertionError` is excluded that coverage, I would have added it as I understood the requirement for making the code defensive, but was not able to figure out how to add coverage for it.

Ok, will update the code to use a same error message for all cases.
